src/extension/gogoduck/src/main/java/au/org/emii/gogoduck/worker/GoGoDuck.java

```diff
@@ -0,0 +1,511 @@
+        logger.info("Enforcing file limit...");
+        if (URIList.size() > limit) {
+            userLog.log("Sorry we cannot process this request due to the amount of files requiring processing.");
+            userLog.log(String.format("The file limit is '%d' and this aggregation job requires '%d' files.", limit, URIList.size()));
+            userLog.log("Please recreate a download request that will require less files to aggregate.");
+            logger.error(String.format("Aggregation asked for %d, we allow only %d", URIList.size(), limit));
+            throw new GoGoDuckException("Too many files");
+        }
+        else if (URIList.size() == 0) {
+            userLog.log("The list of URLs obtained was empty, were your subseting parameters OK?");
+            logger.error("No URLs returned for aggregation");
+            throw new GoGoDuckException("No files returned from geoserver");
+        }
+
+        // All good - keep going :)
+    }
+
+    private void downloadFiles(URIList uriList, Path tmpDir) throws GoGoDuckException {
+        logger.info(String.format("Downloading %d file(s)", uriList.size()));
+
+        for (URI uri : uriList) {
+            if (isCancelled()) {
+                logger.warn("GoGoDuck was cancelled during download.");
+                return;
+            }
+
+            File srcFile = new File(uriList.get(0).toString());
+            String basename = new File(uri.toString()).getName();
+            Path dst = new File(tmpDir + File.separator + basename).toPath();
+
+            if(srcFile.exists() && !srcFile.isDirectory()) {
+                Path src = new File(uri.toString()).toPath();
+
+                try {
+                    logger.info(String.format("Linking '%s' -> '%s'", src, dst));
+                    Files.createSymbolicLink(dst, src);
+                } catch (IOException e) {
+                    userLog.log(String.format("Failed accessing '%s'", src));
+                    throw new GoGoDuckException(e.getMessage());
+                }
+            }
+            else {
+                URL url = fileURItoURL(uri);
+                logger.info(String.format("Downloading '%s' -> '%s'", url.toString(), dst));
+
+                try {
+                    ReadableByteChannel rbc = Channels.newChannel(url.openStream());
+                    FileOutputStream fos = new FileOutputStream(dst.toFile());
+                    fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);
+                }
+                catch (IOException e) {
+                    userLog.log(String.format("Failed downloading '%s'", url));
+                    throw new GoGoDuckException(e.getMessage());
+                }
+            }
+
+            String extension = FilenameUtils.getExtension(dst.getFileName().toString());
+
+            if (extension.equals("gz")) {
+                gunzip(dst.toFile());
+            }
+        }
+    }
+
+    private static URL fileURItoURL(URI uri) {
```

Comment:
Maybe we can have a class which takes care of this ugliness, with all the relevant code there for people to see e.g.:

```java
public class UrlMangler {
  private static final Map<String, String> replacePrefixes ...

  public static URL mangle(URL url) {
    return magic();
  }
}
```